src/compas/artists/robotmodelartist.py

```diff
@@ -88,35 +105,48 @@ def attach_tool_model(self, tool_model):
             The tool that should be attached to the robot's flange.
 
         """
-        self.attached_tool_model = tool_model
-        self.create(tool_model.root, 'attached_tool')
+        self.create(tool_model.root, "attached_tool")
 
         if not tool_model.link_name:
             link = self.model.get_end_effector_link()
             tool_model.link_name = link.name
         else:
             link = self.model.get_link_by_name(tool_model.link_name)
 
+        # don't attach twice on the same link
+        self.attached_tool_models[tool_model.link_name] = tool_model
+
         ee_frame = link.parent_joint.origin.copy()
         initial_transformation = Transformation.from_frame_to_frame(Frame.worldXY(), ee_frame)
 
         sample_geometry = link.collision[0] if link.collision else link.visual[0] if link.visual else None
 
-        if hasattr(sample_geometry, 'current_transformation'):
+        if hasattr(sample_geometry, "current_transformation"):
             relative_transformation = sample_geometry.current_transformation
         else:
             relative_transformation = Transformation()
 
         transformation = relative_transformation.concatenated(initial_transformation)
 
-        self.update_tool(transformation=transformation)
+        self.update_tool(tool=tool_model, transformation=transformation)
 
         tool_model.parent_joint_name = link.parent_joint.name
 
-    def detach_tool_model(self):
-        """Detach the tool.
+    def detach_tool_model(self, tool_model=None):
         """
-        self.attached_tool_model = None
+
```

Comment:
I'd add a summary line here, despite it being rather self-explanatory from the method name, but for consistency.